### DISCUSSION TITLE
Multiple FXGL stages

### DISCUSSION THREAD

**Discussion by s5bug** [AlmasB/FXGL#1273] [Q&A]
Is there any (reasonable) way to run multiple FXGL stages? I have [a current system that works, but only works once](https://github.com/s5bug/chatchatfx/blob/9205cff288e3c3524d724ffaa208642a6b9eb667/src/main/java/tf/bug/chatchatfx/LobbySceneController.java#L78-L87) and I was wondering if there was any trickery I could pull off to make it so that I have access to the various necessary objects on a per-stage level so that multiple instances of my stage could work.

**EDIT** actually, it doesn't entirely work: the scene/viewport dimensions aren't bound to the stage dimensions.

**Comment by AlmasB** [comment]
Hmm, I am not sure -- FXGL certainly expects only one main stage to be active at any given moment, though there are things that can work, depending on what the goal is.

Here's a quick example of having multiple stages that have _some_ objects accessible on a per-stage level, resulting in multiple instances. Effectively, it looks like we are splitting FXGL into two (1. as an engine) which maintains its own "main" stage, while other aspects (2. as a game) exist somewhat independently. So calls to things like `FXGL.getGameWorld()` would now lose their meaning, since there is no single world.

```
import com.almasb.fxgl.app.GameApplication;
import com.almasb.fxgl.app.GameSettings;
import com.almasb.fxgl.app.scene.GameScene;
import com.almasb.fxgl.dsl.FXGL;
import com.almasb.fxgl.entity.GameWorld;
import com.almasb.fxgl.entity.component.Component;
import com.almasb.fxgl.physics.PhysicsWorld;
import javafx.scene.Scene;
import javafx.scene.shape.Rectangle;
import javafx.stage.Stage;
import javafx.util.Duration;

public class MultipleStagesSample extends GameApplication {
    @Override
    protected void initSettings(GameSettings settings) { }

    @Override
    protected void initGame() {
        for (int i = 0; i < 3; i++) {
            FXGL.getExecutor().startAsyncFX(() -> {
                openNewStage();
            });
        }
    }

    private void openNewStage() {
        var gameWorld = new GameWorld();
        var physicsWorld = new PhysicsWorld(600, FXGL.getSettings().getPixelsPerMeter());
        var gameScene = new GameScene(800, 600, gameWorld, physicsWorld);

        // just some random entity
        var e = FXGL.entityBuilder()
                .at(100, 100)
                .view(new Rectangle(50, 50))
                .with(new Component() {
                    @Override
                    public void onUpdate(double tpf) {
                        entity.rotateBy(tpf * 10);
                    }
                })
                .build();

        gameWorld.addEntity(e);

        var timer = FXGL.run(() -> {
            gameScene.step(0.016);
        }, Duration.seconds(0.016));

        var stage = new Stage();
        stage.setWidth(800);
        stage.setHeight(600);
        stage.setScene(new Scene(gameScene.getRoot()));
        stage.setOnHidden(event -> timer.expire());
        stage.show();
    }

    public static void main(String[] args) {
        launch(args);
    }
}
```

Not sure this is of help in any way, but if you provide more details, more than happy to help.

**Reply by s5bug** [reply]
Wow, this looks great! I'll try this out when I next get the time.
1. Does this end up binding the scene size to the stage size, for use with `manualResizeEnabled`?
2. How much do I need that `timer` for? I.e. if I'm not using physics but am still trying to use animated sprites (which I could animate based on `nanoTime` or an `AnimationTimer`?), do I still need it?

**Reply by AlmasB** [reply]
1. No, since any settings applied to the engine only work within the context of the primary stage given to the engine. However, I would expect manually binding it should not create any issues and should be trivial -- just normal JavaFX code.
2. If you are using FXGL-based sprite sheet animations (`AnimatedTexture` / `Texture`), then they are driven by the entity's `ViewComponent`, which in itself is driven by `GameWorld` to which the entity is attached. `GameScene` drives the `GameWorld`, so, yes, you do need the timer. If you are manually animating your sprites, then probably not.
